sympy/functions/combinatorial/factorials.py

```diff
@@ -849,10 +890,83 @@ def eval(cls, n, k):
             from sympy import gamma
             return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
 
+    def _eval_Mod(self, q):
+        n, k = self.args
+        if all(x.is_integer for x in (n, k, q)):
+            if all(x.is_Integer for x in (n, k, q)):
+                n, k = map(int, (n, k))
+                aq = abs(q)
+
+                # handle negative integers k or n
+                if k < 0:
+                    return 0
+                if n < 0:
+                    n = -n + k - 1
+
+                # non negative integers k and n
+                if k > n:
+                    return 0
+
+                isprime = aq.is_prime
+                res, aq = 1, int(aq)
+                if isprime:
+                    if aq < n:
+                        # use Lucas Theorem
+                        N, K = n, k
+                        while N or K:
+                            res = res*binomial(N % aq, K % aq) % aq
+                            N, K = N // aq, K // aq
+
+                    else:
+                        # use factorial Mod
+                        d = n - k
+                        if k > d:
+                            k, d = d, k
+                        kf = 1
+                        for i in range(2, k + 1):
+                            kf = kf*i % aq
+                        df = kf
+                        for i in range(k + 1, d + 1):
+                            df = df*i % aq
+                        res = df
+                        for i in range(d + 1, n + 1):
+                            res = res*i % aq
+
+                        res *= Mod(Pow(kf*df % aq, aq - 2, evaluate=False), aq)
+                        res %= aq
+
+                else:
+                    # use binomial factorization
```

Comment:
This part should preferably be explained in more detail.

Reply:
Which details are to expected, exactly.
Explanation for why explicit factorization is required for generic modulo (since mod inverse might not exist for non-coprime pair)?
or
The details of the implemenation (as in how factorization is carried out)?

Reply:
I would expect an explanation for how the factorization is computed.

Reply:
Okay. Is the following fine?
```
Binomial factorization is performed by calculating the exponents of primes <= n in 
`n! /(k! (n - k)!)`, for non-negative integers n and k.
As the exponent of prime in n! is e_p(n) = [n/p] + [n/p**2] + ...
the exponent of prime in binomial(n, k) would be e_p(n) - e_p(k) - e_p(n - k).
```